### PR TITLE
Fix DataRecord serialization

### LINKED ISSUE #6556: Deserializing record structs doesn't respect default arguments.
Produced on 275.2.0.

## Reproduction

In DataRecordTest.cs, add the following definitions:

```cs
    [DataRecord]
    public partial record struct OneByteOneDefaultIntRecordStruct(byte A, int B = 5);
```

```cs
    [Test]
    public void OneByteOneDefaultIntRecordStructTest()
    {
        var mapping = new MappingDataNode {{"a", "1"}};
        var val = Serialization.Read<OneByteOneDefaultIntRecordStruct>(mapping);

        Assert.Multiple(() =>
        {
            Assert.That(val.A, Is.EqualTo(1));
            Assert.That(val.B, Is.EqualTo(5));
        });
    }
```

Run the tests.  `OneByteOneDefaultIntRecordStructTest` will fail, as `val.B` has a value of 0.

### PR DESCRIPTION
**Pull request by ElectroJr**
DataRecord serialization is pretty broken at the moment. Default values are ignored for structs (see #6556) and simple records using floats or negative default values will cause `SerializationManager.CreateRecordInstantiator()` to generate invalid IL. E.g., neither of these work:
```cs
[DataRecord] public partial record Foo(float A);
[DataRecord] public partial record Bar(int A = -1);
```

I've tried to fix both of these issues, and added a bunch more tests,  but I'm not super confident that I've got all the edge cases and the code is somewhat convoluted. I'm also not that confident in writing IL code, and mostly worked backwards by looking at actual IL that got generated when manually calling record constructors.

Fixes #6556

### CODE REVIEW

File: Robust.Shared/Serialization/Manager/SerializationManager.Instantiation.cs [hunk 1]
```diff
@@ -104,13 +164,13 @@ internal ISerializationManager.InstantiationDelegate<T> GetOrCreateInstantiator<
...
             {
-                CreateValueTypeInstantiator(generator, type);
+                CreateRecordInstantiator(generator, type);
             }
-            else if (isRecord)
+            else if (type.IsValueType)
             {
-                CreateRecordInstantiator(generator, type);
+                CreateValueTypeInstantiator(generator, type);
             }
```
Comment (ElectroJr) [inline Robust.Shared/Serialization/Manager/SerializationManager.Instantiation.cs:114]: Switching these if blocks is what actually fixes #6556. Without this, the constructors of DataRecord structs are basically just ignored.
